### DISCUSSION TITLE
Add the PCF8574 template to the .bin

### DISCUSSION THREAD

**Discussion by DC-555** [arendst/Tasmota#18909] [Q&A]
I would like to add the PCF8574 template to the .bin

The custom template, I need to modify manually later which is not feasible for me, as well as adding the .dat file later.

#undef FALLBACK_MODULE
#undef USER_TEMPLATE
#define USER_TEMPLATE "{\"NAME\":\"Mymodule\",\"GPIO\":[1,1,1,1,640,608,1,1,576,9408,9440,3616,6720,736,5600,1, 0,1,1,5568,0,1,1,1,0,0,0,0,672,704,1,1,1,0,0,1],\"FLAG\":0,\"BASE\" :1,\"CMND\":\"module 0\"}"

//-- Configuration PCF8574 -----------------------
#ifndef USE_PCF8574
#define USE_PCF8574
#endif
#define USE_PCF8574_MODE2
#define USE_PCF8574_SENSOR
#define USE_PCF8574_DISPLAYINPUT
#define USE_PCF8574_MQTTINPUT
#define PCF8574_ADDR1 0x20

**Comment by barbudor** [comment]
For PCF8254 Mode 2, the PCF's template should go in a rul
See how to hardcode rules in your binary here : https://tasmota.github.io/docs/Compile-your-build/#user_rule

**Reply by DC-555** [reply]
I need to use script in my .bin

**Reply by barbudor** [reply]
You may include in your script to create the Dat file if it doesn't exist

**Reply by DC-555** [reply]
I tried to do that, but I wasn't successful.
Another thing I tried was to automatically download the .dmp with:

//#define USER_BACKLOG "WebGetConfig http://192.168.96.24:8000/C049EFF870E4.dmp"
//#define USE_WEBGETCONFIG

but it doesn't work either.

In mode 1 it works, but I have to manually enter and modify the entries, and I don't always have access.
When something goes wrong I would like to be able to just send a reset 1, for example.

**Reply by DC-555** [reply]
The way I tried to add the script was like this:
#define USE_SCRIPT "-y {\"NAME\":\"PCF8574\",\"BASE\":0,\"GPIO\":[160,161,162,163,224,225,0,0]}"  // Add support for script (+17k code)
  #define USE_SCRIPT_FATFS 4                     // Script: Add FAT FileSystem Support
  #define SUPPORT_MQTT_EVENT                     // Support trigger event with MQTT subscriptions (+3k5 code)

**Reply by barbudor** [reply]
> #define USE_SCRIPT "-y {"NAME":"PCF8574","BASE":0,"GPIO":[160,161,162,163,224,225,0,0]}" // Add support for script (+17k code)

Where did you get the idea of doing that ? 
It's wrong
If you want to preload a script inside you benary it should go in `#define PREDEFINED_SCRIPT` and should, as any other script, start with ` >D` 

To create `pcf8574.dat`, that should takes place in the `>B` section of the script
You must first test if the file exist.
If it doesn't, create the file and restart.

**Comment by DC-555** [comment]
and like this too: (but that way it loops at startup)

#define START_SCRIPT_FROM_BOOT
#define USER_BACKLOG "script 1"
#define PRECONFIGURED_SCRIPT ">D -y {\"NAME\":\"PCF8574\",\"BASE\":0,\"GPIO\":[160,161,162,163,224,225,0,0]}"

**Reply by barbudor** [reply]
> #define START_SCRIPT_FROM_BOOT
#define USER_BACKLOG "script 1"

Both lines are doing more or less the same thing

What do you expect `-y`  to do ?
It won't create the `pcf8574.dat`  file
You need to use `>B`  and create the file only if it doesn't exist, then restart

**Reply by DC-555** [reply]
Sorry, could you give me an example?

**Reply by sfromis** [reply]
If you want a file preloaded into the file system, from flashing the ESP32 binary, you can also include a file in the build, as part of the file system which can be created with initial flashing of the `.factory` variant of the binary. No script code needed for that. Example:
https://github.com/arendst/Tasmota/blob/development/platformio_tasmota_env32.ini#L142-L155

**Reply by DC-555** [reply]
Unfortunately it didn't show up in the file manager:

![image](https://github.com/arendst/Tasmota/assets/128267457/53b25d3c-c84e-43ca-ac86-cc3f0089c4cf)

**Reply by sfromis** [reply]
And you **did** flash the `.factory` variant of the binary, not just doing an upgrade?

Note that unless you are building/flashing the zbbrdgpro binary, your modification at that specific point is useless. What I posted was an **example**, not where you "should" do it. It was to illustrate a point of what you can do when building.

**Reply by DC-555** [reply]
I can't load the .factory

Unexpected error: The firmware binary is invalid (magic byte=FF, should be E9)

**Reply by sfromis** [reply]
You likely used a wrong tool to try to do it. Not all tools support this format. For a self-compiled binary. you can use WebSerial ESPTool, or esptool.py. Single file flash at offset 0.
